tests/it/io/json/read.rs

```diff
@@ -33,3 +33,138 @@ fn read_json() -> Result<()> {
+
+    let a_iter = a_iter.into_iter().map(Some);
+    let a_inner = MutableListArray::<i32, MutablePrimitiveArray<f64>>::new_with_field(
+        MutablePrimitiveArray::<f64>::new(),
+        "inner",
+        false,
+    );
+    let mut a_outer =
+        MutableListArray::<i32, MutableListArray<i32, MutablePrimitiveArray<f64>>>::new_with_field(
+            a_inner, "a", false,
+        );
+    a_outer.try_extend(a_iter).unwrap();
+    let a_expected: ListArray<i32> = a_outer.into();
+
+    let b_iter = vec![
+        vec![Some(1), Some(2), Some(3)],
+        vec![Some(7), Some(8), Some(9)],
+    ];
+    let b_iter = b_iter.into_iter().map(Some);
+    let mut b = MutableListArray::<i32, MutablePrimitiveArray<i64>>::new_with_field(
+        MutablePrimitiveArray::<i64>::new(),
+        "b",
+        false,
+    );
+    b.try_extend(b_iter).unwrap();
+    let b_expected: ListArray<i32> = b.into();
+
+    let json = json_deserializer::parse(data)?;
+
+    let schema = read::infer_records_schema(&json)?;
+    let actual = read::deserialize_records(&json, &schema)?;
+
+    for (f, arr) in schema.fields.iter().zip(actual.arrays().iter()) {
+        let (expected, actual) = if f.name == "a" {
+            (&a_expected, arr.as_ref())
+        } else if f.name == "b" {
+            (&b_expected, arr.as_ref())
+        } else {
+            panic!("unexpected field found: {}", f.name);
+        };
+
+        // No idea why assert_eq! doesn't work here, but this does.
+        assert_eq!(format!("{:?}", expected), format!("{:?}", actual));
+    }
+
+    Ok(())
+}
+
+#[test]
+fn read_json_fixed_size_records() -> Result<()> {
+    let data = br#"[
+        {
+            "a": [1, 2.2, 3, 4]
+        },
+        {
+            "a": [5, 6, 7, 8]
+        },
+        {
+            "a": [7, 8, 9]
+        }
+    ]"#;
+
+    let a_iter = vec![
+        Some(vec![Some(1.), Some(2.2), Some(3.), Some(4.)]),
+        Some(vec![Some(5.), Some(6.), Some(7.), Some(8.)]),
+        None,
+    ];
+
+    let a_iter = a_iter.into_iter();
+    let mut a = MutableFixedSizeListArray::<MutablePrimitiveArray<f64>>::new_with_field(
+        MutablePrimitiveArray::<f64>::new(),
+        "inner",
+        false,
+        4,
+    );
+    a.try_extend(a_iter).unwrap();
+    let a_expected: FixedSizeListArray = a.into();
+
+    let json = json_deserializer::parse(data)?;
+
+    let schema = Schema {
+        fields: vec![Field::new("a", a_expected.data_type().clone(), true)],
+        metadata: Metadata::default(),
+    };
+    let actual = read::deserialize_records(&json, &schema)?;
+
+    for (f, arr) in schema.fields.iter().zip(actual.arrays().iter()) {
+        let (expected, actual) = if f.name == "a" {
+            (&a_expected, arr.as_ref())
+        } else {
+            panic!("unexpected field found: {}", f.name);
+        };
+
+        // No idea why assert_eq! doesn't work here, but this does.
+        assert_eq!(format!("{:?}", expected), format!("{:?}", actual));
```

Comment:
I suspect that this is because line 144 uses `false` on the nullability of the inner field, but the result is a nullability true

Reply:
Yup, this was 100% it. Kinda an interesting philosophical question on whether they are actually equal... value-wise, they are identical, even though one _could_ hold nullable values. While I'm not 100% sure I agree with the conclusion, I understand and respect the reasoning.

Reply:
The last thought I'll leave here is that the error when they are not equal is pretty confusing:

```
thread 'io::json::read::read_json_records' panicked at 'assertion failed: `(left == right)`
  left: `ListArray[[[1.1, 2, 3], [2, 3], [4, 5, 6]], [[3, 2, 1], [3, 2], [6, 5, 4]]]`,
 right: `ListArray[[[1.1, 2, 3], [2, 3], [4, 5, 6]], [[3, 2, 1], [3, 2], [6, 5, 4]]]`', tests/it/io/json/read.rs:114:9
```

Not sure if the solution is to somehow mark that one is nullable and one is not.

Reply:
Yup, that makes sense. Especially with deeply nested types.